### ISSUE TITLE
Issue involving Never type in pattern match

### ISSUE THREAD

**Issue by mmcdon20** [dart-lang/language#2953]
`Dart SDK version: 3.0.0-366.0.dev (dev) (Fri Mar 24 06:44:03 2023 -0700) on "windows_x64"`

Consider this program.

```dart
void main() {}

LinkedList<T> head<T>(LinkedList<T> list) => switch (list) {
      Nil() => Nil(),
      Cons(first: var x, rest: _) => Cons(x, Nil()),
    };

sealed class LinkedList<T> {}

class Nil extends LinkedList<Never> {}

class Cons<T> extends LinkedList<T> {
  Cons(this.first, this.rest);
  final T first;
  final LinkedList<T> rest;
}
```

This program produces the following compile error.

```
bin/experiment.dart:3:46: Error: A value of type 'Object' can't be returned from a function with return 
type 'LinkedList<T>'.
 - 'Object' is from 'dart:core'.
 - 'LinkedList' is from 'bin/experiment.dart'.
LinkedList<T> head<T>(LinkedList<T> list) => switch (list) {
                                             ^
```

The program will compile if you add a typecast in the `head` function.

```dart
LinkedList<T> head<T>(LinkedList<T> list) => switch (list) {
      Nil() => Nil() as LinkedList<T>,
      Cons(first: var x, rest: _) => Cons(x, Nil()),
    };
```

The above compiles and runs but it also produces an analyzer [unnecessary_cast](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=unnecessary_cast#unnecessary_cast) message.

Unless I am misunderstanding something, I believe this should compile without the need for a cast since `Never` is a bottom type and therefore a `LinkedList<Never>` is also a `LinkedList<T>`.

**Comment by leafpetersen** [comment]
> You could say that we should just type check each branch separately: `Nil() <: LinkedList<T>` is true, so we can return the value from the first branch just fine, and similarly for the second branch. However, this implies that the switch expression would not have a type of its own,

This isn't really true, I think.  [This proposed change](https://github.com/dart-lang/language/issues/2129) would solve this particular issue: we would say that when we have a context type, we give that type to the switch expression, and then check each arm against that type.  We would only use upper bound when there is no context.  I somewhat suspect that switch expressions are going to make this problem much worse, and that we will likely want to adopt #2129 soonish.  cc @stereotype441 

Concretely to the example code, making `Nil` generic instead of monomorphic will allow inference + upper bound to do the right thing here.  Not entirely satisfying, I realize, but pragmatically probably the best solution at the moment.

```dart
class Nil<T> extends LinkedList<T> {}
```

**Comment by munificent** [comment]
I agree with @leafpetersen. I think LUB is almost never what you want, and if there is a context type that can be pushed down, it's better to do that.

At the same time, it's probably too close to the wire to make this change to switch expressions for Dart 3.0, so we may have to tweak it in a future release.

**Comment by eernstg** [comment]
@leafpetersen wrote:
> when we have a context type, we give that type to the switch expression, and then check each arm against that type.

Ah, that's a nice trick!

I guess we would type check each case using the given context type schema `S`, and that would yield a set of types `T1 .. Tk`, one for each case, and then we'd use `T1 .. Tk` to fill in the `_` in `S`, yielding a type `T` which would then be the type of the switch/conditional/etc. (That would be a more precise type than we would get from turning `S` into a type immediately, and then type checking each case against that). Finally, we might need to check `Tj <: T` for each `j`, if that isn't already known to be true by construction.

Presumably this approach would be used with conditional expressions as well.

The **UP** function is used in several other situations, but they give rise to the introduction of a separate entity whose type inference will already allow the context type schema to be taken into account (which means that they already behave in a way which is very similar to the proposed rule for switches):

```dart
abstract class LinkedList<X> {}
class Nil extends LinkedList<Never> {}
class Cons<X> extends LinkedList<X> {}

void main() {
  // Inferred type of `f`: `Object Function()`. Also accepted with declared type.
  /*LinkedList<int> Function()*/ var f = () {
    if (2 > 1) return Nil();
    return Cons();
  };

  // Inferred: `List<Object>`, `Set<Object>`, `Map<Object, Null>`, also ok with declared type.
  /*List<LinkedList<int>>*/ var l = [Nil(), Cons()];
  /*Set<LinkedList<int>>*/ var s = {Nil(), Cons()};
  /*Map<LinkedList<int>, Null>*/ var m = {Nil(): null, Cons(): null};
}
```

It makes a lot of sense that the switch expression should be equally responsive to the context type schema as the function literals and collection literals. One way to build an intuition about this could be that the switch is considered to be a function-like entity, except that it is always invoked immediately (similarly to expression blocks).

Nevertheless, this also illustrates that we may still need to adjust **UP**, because the `Object` part of the inferred types in the example may still create difficulties further down in the code when the declared entities are used.

**Comment by mmcdon20** [comment]
This appears to have been resolved by #1618. I assume some of the other issues under [least-upper-bound](https://github.com/dart-lang/language/issues?q=is%3Aopen+is%3Aissue+label%3Aleast-upper-bound) have been resolved by this as well.
